File: backend/src/modules/userdata/userdata.service.ts
import { Injectable, NotFoundException } from '@nestjs/common';
import { InjectRepository } from '@nestjs/typeorm';
import { Repository, Like } from 'typeorm';
import { DbAliases } from './entities/dbaliases.entity';
import { Grp } from './entities/grp.entity';
import { SpeedDial } from './entities/speed-dial.entity';
import { PaginationDto, PaginatedResult } from '../../common/dto/pagination.dto';

@Injectable()
export class UserdataService {
  constructor(
    @InjectRepository(DbAliases) private aliasRepository: Repository<DbAliases>,
    @InjectRepository(Grp) private grpRepository: Repository<Grp>,
    @InjectRepository(SpeedDial) private speedDialRepository: Repository<SpeedDial>,
  ) {}

  // ========== Aliases ==========
  async findAllAliases(dto: PaginationDto): Promise<PaginatedResult<DbAliases>> {
    const { page, limit, keyword } = dto;
    const skip = (page - 1) * limit;
    const where: any = {};
    if (keyword) where.aliasUsername = Like(`%${keyword}%`);
    const [items, total] = await this.aliasRepository.findAndCount({ where, skip, take: limit, order: { id: 'ASC' } });
    return new PaginatedResult(items, total, page, limit);
  }

  async createAlias(data: Partial<DbAliases>): Promise<DbAliases> {
    return this.aliasRepository.save(this.aliasRepository.create(data));
  }

  async updateAlias(id: number, data: Partial<DbAliases>): Promise<DbAliases> {
    const item = await this.aliasRepository.findOne({ where: { id } });
    if (!item) throw new NotFoundException('别名不存在');
    Object.assign(item, data);
    return this.aliasRepository.save(item);
  }

  async removeAlias(id: number): Promise<void> {
    await this.aliasRepository.delete(id);
  }

  // ========== Groups ==========
  async findAllGroups(dto: PaginationDto): Promise<PaginatedResult<Grp>> {
    const { page, limit, keyword } = dto;
    const skip = (page - 1) * limit;
    const where: any = {};
    if (keyword) where.username = Like(`%${keyword}%`);
    const [items, total] = await this.grpRepository.findAndCount({ where, skip, take: limit, order: { id: 'ASC' } });
    return new PaginatedResult(items, total, page, limit);
  }

  async createGroup(data: Partial<Grp>): Promise<Grp> {
    return this.grpRepository.save(this.grpRepository.create(data));
  }

  async updateGroup(id: number, data: Partial<Grp>): Promise<Grp> {
    const item = await this.grpRepository.findOne({ where: { id } });
    if (!item) throw new NotFoundException('分组不存在');
    Object.assign(item, data);
    return this.grpRepository.save(item);
  }

  async removeGroup(id: number): Promise<void> {
    await this.grpRepository.delete(id);
  }

  // ========== Speed Dial ==========
  async findAllSpeedDials(dto: PaginationDto): Promise<PaginatedResult<SpeedDial>> {
    const { page, limit, keyword } = dto;
    const skip = (page - 1) * limit;
    const where: any = {};
    if (keyword) where.username = Like(`%${keyword}%`);
    const [items, total] = await this.speedDialRepository.findAndCount({ where, skip, take: limit, order: { id: 'ASC' } });
    return new PaginatedResult(items, total, page, limit);
  }

  async createSpeedDial(data: Partial<SpeedDial>): Promise<SpeedDial> {
    return this.speedDialRepository.save(this.speedDialRepository.create(data));
  }

  async updateSpeedDial(id: number, data: Partial<SpeedDial>): Promise<SpeedDial> {
    const item = await this.speedDialRepository.findOne({ where: { id } });
    if (!item) throw new NotFoundException('快捷拨号不存在');
    Object.assign(item, data);
    return this.speedDialRepository.save(item);
  }

  async removeSpeedDial(id: number): Promise<void> {
    await this.speedDialRepository.delete(id);
  }
}

File: kam-admin-module-user/src/main/java/io/kamailio/admin/modules/userdata/service/UserdataService.java
package io.kamailio.admin.modules.userdata.service;

import io.kamailio.admin.common.dto.PaginatedResult;
import io.kamailio.admin.common.dto.PaginationDto;
import io.kamailio.admin.modules.userdata.entity.DbAliases;
import io.kamailio.admin.modules.userdata.entity.Grp;
import io.kamailio.admin.modules.userdata.entity.SpeedDial;
import io.kamailio.admin.modules.userdata.repository.DbAliasesRepository;
import io.kamailio.admin.modules.userdata.repository.GrpRepository;
import io.kamailio.admin.modules.userdata.repository.SpeedDialRepository;
import org.springframework.data.domain.PageRequest;
import org.springframework.data.domain.Sort;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

@Service
public class UserdataService {
    private final DbAliasesRepository aliasRepository;
    private final GrpRepository grpRepository;
    private final SpeedDialRepository speedDialRepository;

    public UserdataService(DbAliasesRepository aliasRepository, GrpRepository grpRepository, SpeedDialRepository speedDialRepository) {
        this.aliasRepository = aliasRepository;
        this.grpRepository = grpRepository;
        this.speedDialRepository = speedDialRepository;
    }

    public PaginatedResult<DbAliases> findAllAliases(PaginationDto dto) {
        var page = dto.getKeyword() != null && !dto.getKeyword().isBlank()
                ? aliasRepository.findByAliasUsernameContaining(dto.getKeyword(), PageRequest.of(dto.pageOrDefault() - 1, dto.limitOrDefault(), Sort.by("id")))
                : aliasRepository.findAll(PageRequest.of(dto.pageOrDefault() - 1, dto.limitOrDefault(), Sort.by("id")));
        return new PaginatedResult<>(page.getContent(), page.getTotalElements(), dto.pageOrDefault(), dto.limitOrDefault());
    }

    @Transactional
    public DbAliases createAlias(DbAliases data) { return aliasRepository.save(data); }

    @Transactional
    public DbAliases updateAlias(Integer id, DbAliases data) {
        var item = aliasRepository.findById(id).orElseThrow(() -> new RuntimeException("别名不存在"));
        if (data.getAliasUsername() != null) item.setAliasUsername(data.getAliasUsername());
        if (data.getAliasDomain() != null) item.setAliasDomain(data.getAliasDomain());
        if (data.getUsername() != null) item.setUsername(data.getUsername());
        if (data.getDomain() != null) item.setDomain(data.getDomain());
        return aliasRepository.save(item);
    }

    @Transactional
    public void removeAlias(Integer id) { aliasRepository.deleteById(id); }

    public PaginatedResult<Grp> findAllGroups(PaginationDto dto) {
        var page = dto.getKeyword() != null && !dto.getKeyword().isBlank()
                ? grpRepository.findByUsernameContaining(dto.getKeyword(), PageRequest.of(dto.pageOrDefault() - 1, dto.limitOrDefault(), Sort.by("id")))
                : grpRepository.findAll(PageRequest.of(dto.pageOrDefault() - 1, dto.limitOrDefault(), Sort.by("id")));
        return new PaginatedResult<>(page.getContent(), page.getTotalElements(), dto.pageOrDefault(), dto.limitOrDefault());
    }

    @Transactional
    public Grp createGroup(Grp data) { return grpRepository.save(data); }

    @Transactional
    public Grp updateGroup(Integer id, Grp data) {
        var item = grpRepository.findById(id).orElseThrow(() -> new RuntimeException("分组不存在"));
        if (data.getUsername() != null) item.setUsername(data.getUsername());
        if (data.getDomain() != null) item.setDomain(data.getDomain());
        if (data.getGrp() != null) item.setGrp(data.getGrp());
        return grpRepository.save(item);
    }

    @Transactional
    public void removeGroup(Integer id) { grpRepository.deleteById(id); }

    public PaginatedResult<SpeedDial> findAllSpeedDials(PaginationDto dto) {
        var page = dto.getKeyword() != null && !dto.getKeyword().isBlank()
                ? speedDialRepository.findByUsernameContaining(dto.getKeyword(), PageRequest.of(dto.pageOrDefault() - 1, dto.limitOrDefault(), Sort.by("id")))
                : speedDialRepository.findAll(PageRequest.of(dto.pageOrDefault() - 1, dto.limitOrDefault(), Sort.by("id")));
        return new PaginatedResult<>(page.getContent(), page.getTotalElements(), dto.pageOrDefault(), dto.limitOrDefault());
    }

    @Transactional
    public SpeedDial createSpeedDial(SpeedDial data) { return speedDialRepository.save(data); }

    @Transactional
    public SpeedDial updateSpeedDial(Integer id, SpeedDial data) {
        var item = speedDialRepository.findById(id).orElseThrow(() -> new RuntimeException("快捷拨号不存在"));
        if (data.getUsername() != null) item.setUsername(data.getUsername());
        if (data.getDomain() != null) item.setDomain(data.getDomain());
        if (data.getSdUsername() != null) item.setSdUsername(data.getSdUsername());
        if (data.getSdDomain() != null) item.setSdDomain(data.getSdDomain());
        if (data.getNewUri() != null) item.setNewUri(data.getNewUri());
        if (data.getDescription() != null) item.setDescription(data.getDescription());
        return speedDialRepository.save(item);
    }

    @Transactional
    public void removeSpeedDial(Integer id) { speedDialRepository.deleteById(id); }
}
